feat: add support for filtering apps by path and files

slashexx opened the pull request:
<!--
Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.
-->
Fixes #21052 

Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [ ] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [ ] Optional. My organization is added to USERS.md.
* [ ] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

<!-- Please see [Contribution FAQs](https://argo-cd.readthedocs.io/en/latest/developer-guide/faq/) if you have questions about your pull-request. -->

nitishfy reviewed (commented):
Is this PR ready for review? Also, can you fix the CI checks?

slashexx commented:
@nitishfy Thanks for following up, apologies for the delay in fixing this.

The CI checks should now work fine, I've also marked this PR as ready for review, PTAL !

nitishfy commented on util/argo/argo.go:
```diff
@@ -115,6 +115,43 @@ func FilterByProjects(apps []argoappv1.Application, projects []string) []argoapp
+func FilterByPath(apps []argoappv1.Application, path string) []argoappv1.Application {
+	filteredApps := make([]argoappv1.Application, 0)
+	for _, app := range apps {
+		if app.Spec.Source != nil && (app.Spec.Source.Path == path || strings.HasPrefix(app.Spec.Source.Path, path+"/")) {
```
```suggestion
		if app.Spec.Source != nil && (app.Spec.Source.Path == path || strings.HasPrefix(app.Spec.Source.Path, path+"/")) {
```
You'd also need to filter the apps that consist of multiple sources. Right now, it only checks for a single source application.

nitishfy replied:
```suggestion
		if app.Spec.Source != nil && (app.Spec.Source.Path == path || strings.HasPrefix(app.Spec.Source.Path, path+"/")) {
```
Checking for a path is simply not sufficient. You should also check for a absolute/relative path too. For eg. in any  setup, I should be simply able to pass the file path to a file that exists locally and it should internally check if any application is using that specified path for sync operation.  

@agaudreault Isn't this what we plan to do? Otherwise with the current setup we just pass an argument to the path flag and check if that matches with the source path.

nitishfy commented on util/argo/argo.go:
```diff
@@ -115,6 +115,43 @@ func FilterByProjects(apps []argoappv1.Application, projects []string) []argoapp
+func FilterByPath(apps []argoappv1.Application, path string) []argoappv1.Application {
+	filteredApps := make([]argoappv1.Application, 0)
+	for _, app := range apps {
+		if app.Spec.Source != nil && (app.Spec.Source.Path == path || strings.HasPrefix(app.Spec.Source.Path, path+"/")) {
```
```suggestion
		if app.Spec.Source != nil && (app.Spec.Source.Path == path || strings.HasPrefix(app.Spec.Source.Path, path+"/")) {
```
why are you checking for `strings.HasPrefix()` here?

slashexx replied:
Hi @nitishfy , apologies I'm a bit new around here.

I've put it like that so it compares the exact path along with subdirectories and all ?

Is it not required ?

nitishfy replied:
You'd need to check for absolute and relative path here instead of just checking if a path consist of `/`. Isn't this the intended behaviour? cc @agaudreault

nitishfy reviewed (changes requested):
Thanks for working on this. I've not reviewed the `--file` flag implementation yet. However, i think a lot of things can be improved for the `--path` flag implementation.

slashexx commented:
> Thanks for working on this. I've not reviewed the `--file` flag implementation yet. However, i think a lot of things can be improved for the `--path` flag implementation.

Thanks for reviewing this PR, I'll make the required changes and change --file flag accordingly as well.

nitishfy reviewed (commented):
I'm yet to test these changes again. Meanwhile, can you fix the failing tests?

slashexx commented:
@nitishfy apologies for the huge delay, was a little occupied with stuff.

I've fixed the failing tests and refactored the path function to use absolute pathing system, ptal !
